File: Engine/src/main/java/org/etieskrill/engine/graphics/texture/animation/AnimatedTexture.java
package org.etieskrill.engine.graphics.texture.animation;

import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import lombok.experimental.Delegate;
import org.etieskrill.engine.graphics.texture.AbstractTexture.MagFilter;
import org.etieskrill.engine.graphics.texture.AbstractTexture.MinFilter;
import org.etieskrill.engine.graphics.texture.AbstractTexture.Wrapping;
import org.etieskrill.engine.graphics.texture.ArrayTexture;
import org.etieskrill.engine.graphics.texture.Texture2D;
import org.joml.Vector2ic;

import java.nio.ByteBuffer;
import java.util.List;

import static lombok.AccessLevel.NONE;
import static org.etieskrill.engine.graphics.texture.animation.TextureAnimationYamlParser.loadAnimationMetadata;
import static org.etieskrill.engine.util.FileUtils.splitTypeFromPath;
import static org.lwjgl.BufferUtils.createByteBuffer;
import static org.lwjgl.BufferUtils.zeroBuffer;

@Getter
public class AnimatedTexture {

    private final @Delegate ArrayTexture texture;
    private final TextureAnimationMetadata metaData;

    public static AnimatedTextureBuilder builder() {
        return new AnimatedTextureBuilder();
    }

    @Setter
    @Accessors(fluent = true)
    public static final class AnimatedTextureBuilder {
        private static final String META_FILE_EXTENSION = ".tex-anim.yml";

        private String file;
        private String metaFile;

        //TODO much yikes, very eww
        private @Setter(NONE) MinFilter minFilter = MinFilter.TRILINEAR;
        private @Setter(NONE) MagFilter magFilter = MagFilter.LINEAR;
        private @Setter(NONE) Wrapping wrapping = Wrapping.REPEAT;

        public AnimatedTextureBuilder setMipMapping(MinFilter minFilter, MagFilter magFilter) {
            this.minFilter = minFilter;
            this.magFilter = magFilter;
            return this;
        }

        public AnimatedTextureBuilder setWrapping(Wrapping wrapping) {
            this.wrapping = wrapping;
            return this;
        }

        public AnimatedTexture build() {
            if (file == null) throw new NullPointerException("Texture file may not be null");
            if (metaFile == null) {
                var typedFile = splitTypeFromPath(file);
                metaFile = typedFile.getFileName() + META_FILE_EXTENSION;
            }

            return new AnimatedTexture(this);
        }
    }

    private AnimatedTexture(AnimatedTextureBuilder builder) {
        var textureAtlas = new Texture2D.FileBuilder(builder.file);
        metaData = loadAnimationMetadata(builder.metaFile);

        if (metaData.getFrameSize().x() * metaData.getFrameSize().y() * metaData.getFrames().size()
                != textureAtlas.getTextureData().capacity() / textureAtlas.getFormat().getChannels()) {
            throw new IllegalArgumentException("Metadata size does not match texture size");
        }

        var animatedTexture = new ArrayTexture.BufferBuilder(
                metaData.getFrameSize(),
                metaData.getFrames().size(),
                metaData.getFormat() != null ? metaData.getFormat() : textureAtlas.getFormat());

        animatedTexture.setMipMapping(builder.minFilter, builder.magFilter);
        animatedTexture.setWrapping(builder.wrapping);

        addTextures(textureAtlas, animatedTexture);
        texture = animatedTexture.build();
    }

    private void addTextures(Texture2D.Builder textureAtlas, ArrayTexture.BufferBuilder animatedTexture) {
        Vector2ic textureSize = textureAtlas.getPixelSize();
        int pixelChannels = textureAtlas.getFormat().getChannels();

        ByteBuffer frameTexture = createByteBuffer(metaData.getFrameSize().x() * metaData.getFrameSize().y() * pixelChannels);
        List<TextureAnimationFrame> frames = metaData.getFrames();
        for (int i = 0; i < frames.size(); i++) {
            TextureAnimationFrame frame = frames.get(i);

            frameTexture.rewind();
            zeroBuffer(frameTexture);

            int width = frame.getAtlasArea().lengthX(), height = frame.getAtlasArea().lengthY();
            for (int h = 0; h < height; h++) {
                frameTexture.put(
                        width * h * pixelChannels,
                        textureAtlas.getTextureData(),
                        (textureSize.x() * h + width * i) * pixelChannels,
                        width * pixelChannels
                );
                frameTexture.position(frameTexture.position() + width * pixelChannels);
            }

            animatedTexture.addTexture(frameTexture.rewind());
        }
    }

}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/graphics/texture/animation/AnimatedTexture.kt
package org.etieskrill.engine.graphics.texture.animation

import org.etieskrill.engine.graphics.texture.ArrayTexture2D
import org.etieskrill.engine.graphics.texture.TextureData
import org.etieskrill.engine.graphics.texture.TextureMagFilter
import org.etieskrill.engine.graphics.texture.TextureMinFilter
import org.etieskrill.engine.graphics.texture.TextureType
import org.etieskrill.engine.graphics.texture.TextureWrapping
import org.etieskrill.engine.graphics.texture.loadTexture2DData
import org.etieskrill.engine.util.extension
import org.etieskrill.engine.util.pathName
import org.etieskrill.engine.util.subExtension
import org.joml.Vector2ic

class AnimatedTexture(
    file: String,
    metaFile: String = file.pathName + META_FILE_EXTENSION,
    minFilter: TextureMinFilter = TextureMinFilter.TRILINEAR,
    magFilter: TextureMagFilter = TextureMagFilter.LINEAR,
    wrapping: TextureWrapping = TextureWrapping.REPEAT
) {

    companion object {
        private const val META_FILE_EXTENSION = ".tex-anim.yml"
    }

    val texture: ArrayTexture2D
    val metadata: TextureAnimationMetadata
    val pixelSize get() = texture.size
    val length get() = texture.length

    init {
        check(metaFile.subExtension == "tex-anim" && metaFile.extension == "yml") {
            "Meta file must have 'tex-anim.yml' extension"
        }

        val textureData = loadTexture2DData(file, TextureType.DIFFUSE)
        metadata = loadAnimationMetadata(metaFile, textureData.format)

        check(
            metadata.frameSize.x() * metadata.frameSize.y() * metadata.frames.size
                    != textureData.buffer.size / textureData.format.numChannels
        ) { "Metadat size does not match texture size" }

        texture = ArrayTexture2D(
            metadata.frameSize,
            metadata.frames.size,
            buffer = unpackTextureData(metadata, textureData),
            type = TextureType.DIFFUSE,
            format = metadata.format,
            minFilter = minFilter,
            magFilter = magFilter,
            wrapping = wrapping,
        )
    }

    //FIXME this only works when each frame is the exact same size, right? TODO add padding if required
    private fun unpackTextureData(metadata: TextureAnimationMetadata, data: TextureData): ByteArray {
        val arrayData = ByteArray(data.buffer.size)
        var arrayHead = 0
        val pixelChannels = data.format.numChannels

        metadata.frames.forEachIndexed { i, frame ->
            val width = frame.atlasArea.lengthX()
            val height = frame.atlasArea.lengthY()
            for (h in 0 until height) {
                val sourceIndex = data.size.x() * h + width * i
                data.buffer.copyInto(
                    arrayData,
                    arrayHead,
                    sourceIndex * pixelChannels,
                    (sourceIndex + width) * pixelChannels
                )
                arrayHead += width * pixelChannels
            }
        }

        return arrayData
    }

}
